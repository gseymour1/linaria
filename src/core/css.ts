type CSSMeta = string;

type CSSProperties = {
  [key: string]: string | number | CSSProperties;
};

export default function css(
  _strings: TemplateStringsArray,
  ..._exprs: Array<string | number | CSSProperties | CSSMeta>
): string {
  throw new Error(
    'Using the "css" tag in runtime is not supported. Make sure you have set up the Babel plugin correctly.'
  );
}
