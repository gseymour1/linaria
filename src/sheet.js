/* @flow */

import stylis from 'stylis';

function sheet() {
  let ruleCache: ?Array<CSSRule>;
  let cssText: ?string;
  let stylesCache: ?Array<{ selector: string, css: string }>;

  if (typeof window === 'undefined' || typeof window.document === 'undefined') {
    return {
      insert(selector: string, css: string) {
        const text = stylis(selector, css);
        stylesCache = stylesCache || [];
        stylesCache.push({ selector, css });
        cssText = cssText ? cssText + text : text;
      },
      rules() {
        throw new Error('Not implemented');
      },
      styles() {
        return stylesCache || [];
      },
      dump() {
        const result = cssText || '';
        cssText = null;
        stylesCache = null;
        return result;
      },
    };
  }

  const style = document.createElement('style');
  const node = document.createTextNode('');

  style.appendChild(node);
  style.setAttribute('type', 'text/css');

  if (document.head != null) {
    document.head.appendChild(style);
  } else {
    throw new Error('Unable to insert stylesheet');
  }

  return {
    insert(selector: string, css: string) {
      const text = stylis(selector, css);
      node.appendData(text);
      // invalidate the cache since stylesheets have changed
      ruleCache = null;
    },
    rules() {
      if (ruleCache != null) {
        return { list: ruleCache, cache: true };
      }

      ruleCache = [].concat(
        /* $FlowFixMe */
        ...Array.from(document.styleSheets).map(s => Array.from(s.cssRules))
      );

      return { list: ruleCache, cache: false };
    },
  };
}

export default sheet();
