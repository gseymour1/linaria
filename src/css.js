/* @flow */

import slugify from './slugify';
import sheet from './sheet';

const named = (name?: string = 'css') => (
  template: string[],
  ...expressions: string[]
) => {
  const styles = template.reduce(
    (accumulator, part, i) => accumulator + expressions[i - 1] + part
  );

  const slug = slugify(styles);
  const classname = `${name}_${slug}`;

  sheet.append(`.${classname}`, styles);

  return classname;
};

const css = named();

css.named = named;

export default css;
