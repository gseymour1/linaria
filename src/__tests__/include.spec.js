/* @flow */

import sheet from '../sheet';
import css from '../css';
import include from '../include';

describe('include module', () => {
  it('should include styles from a single class name', () => {
    const text = css`font-weight: 400;`;

    const title = css`
      color: orange;

      @media (max-width: 100px) {
        font-family: monospace;
      }
    `;

    const header = css`
      ${include(text, title)};
      font-family: sans-serif;
    `;

    expect(include(title)).toMatchSnapshot();
    expect(include(text, title)).toMatchSnapshot();
    expect(sheet.styles()[`.${header}`]).toMatchSnapshot();
    expect(sheet.dump()).toMatchSnapshot();
  });

  it('should throw for invalid class name', () => {
    expect(() => include('not_valid')).toThrow(
      `Unable to find CSS for the class name: not_valid`
    );
  });
});
