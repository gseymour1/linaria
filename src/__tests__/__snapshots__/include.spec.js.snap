// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`include module should include styles from a single class name 1`] = `
"
color: orange;

@media (max-width: 100px) {
  font-family: monospace;
}"
`;

exports[`include module should include styles from a single class name 2`] = `
"
font-weight: 400;
color: orange;

@media (max-width: 100px) {
  font-family: monospace;
}"
`;

exports[`include module should include styles from a single class name 3`] = `
"font-weight: 400;
color: orange;

@media (max-width: 100px) {
font-family: monospace;
};
    font-family: sans-serif;"
`;

exports[`include module should include styles from a single class name 4`] = `
".css__n32n50{font-weight:400;}
.css__1adt32q{color:orange;}@media (max-width:100px){.css__1adt32q{font-family:monospace;}}
.css__1xk7l0x{font-weight:400;color:orange;font-family:sans-serif;}@media (max-width:100px){.css__1xk7l0x{font-family:monospace;}}"
`;
